test/unit/port.cc

```diff
@@ -0,0 +1,779 @@
+#include "lua_test_utils.h"
+
+#include "box/lua/call.h"
+#include "box/lua/tuple.h"
+#include "box/port.h"
+#include "box/session.h"
+#include "box/tuple.h"
+#include "box/user.h"
+#include "core/event.h"
+#include "core/port.h"
+#include "lua/init.h"
+#include "lua/minifio.h"
+#include "lua/msgpack.h"
+#include "lua/utils.h"
+#include "memory.h"
+#include "msgpuck.h"
+#include "small/obuf.h"
+#include "trivia/util.h"
+
+#define UNIT_TAP_COMPATIBLE 1
+#include "unit.h"
+
+/**
+ * STRUCTURE OF THE TEST
+ *
+ * The test consists of several helper sections (e.g. Lua helpers,
+ * MsgPack helpers) and several test sections. Each test section
+ * focuses on a particular port implementation.
+ *
+ * A test section has a function that creates port and fills it with fixed
+ * data. Then, every test case, dedicated to a particular port method or group
+ * of methods, checks if the empty port works correctly, then fills the
+ * port with the filler function and manually creates an expected
+ * MsgPack packet/Lua stack/etc and passes the port and the expected
+ * object to checker.
+ *
+ * Method port_dump_plain is not tested because it requires console.lua file,
+ * so it's easier to test it from Lua then loading the Lua file in unit test.
+ */
+
+/**
+ * This helper is placed here to be used in Lua helpers.
+ * See actual description near the definition.
+ */
+static void
+test_check_mp_equal(const char *got, uint32_t got_size,
+		    const char *expected, uint32_t expected_size,
+		    bool no_header);
+
+/**
+ * Utils to check port_dump_lua method.
+ *
+ * Checker for port_dump_lua with mp_object mode is in msgpack
+ * helpers section because it is actually dump_msgpack method that
+ * pushes its result to Lua.
+ */
+
+/**
+ * Returns true if two tables on top of passed Lua stack are equal,
+ * false is returned otherwise. Compared tables are popped.
+ */
+static bool
+lua_table_equal(struct lua_State *L)
+{
+	const char *text = "function table_eq(a, b) "
+		"if type(a) ~= 'table' or type(b) ~= 'table' then "
+		"    return a == b "
+		"end "
+		"for k, v in pairs(a) do "
+		"    if not table_eq(v, b[k]) then "
+		"        return false "
+		"    end "
+		"end "
+		"for k, _ in pairs(b) do "
+		"    if type(a[k]) == 'nil' then "
+		"        return false "
+		"    end "
+		"end "
+		"return true "
+		"end "
+		"return table_eq";
+	int rc = luaT_dostring(L, text);
+	fail_if(rc != 0);
```

Comment:
Nit: you could cache the function definition in `main` once.

Reply:
Done